pkg/resources/management.cattle.io/v3/project/validator.go

```diff
@@ -0,0 +1,211 @@
+package project
+
+import (
+	"fmt"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/webhook/pkg/admission"
+	objectsv3 "github.com/rancher/webhook/pkg/generated/objects/management.cattle.io/v3"
+	"github.com/rancher/wrangler/pkg/data/convert"
+	corev1controllers "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/kubelet/util/format"
+	"k8s.io/utils/trace"
+)
+
+const (
+	systemProjectLabel = "authz.management.cattle.io/system-project"
+)
+
+// Validator implements admission.ValidatingAdmissionWebhook.
+type Validator struct {
+	admitter admitter
+}
+
+// NewValidator returns a project validator.
+func NewValidator(namespaceCache corev1controllers.NamespaceCache) *Validator {
+	return &Validator{
+		admitter{
+			namespaceCache: namespaceCache,
+		},
+	}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{
+		admissionregistrationv1.Create,
+		admissionregistrationv1.Update,
+		admissionregistrationv1.Delete,
+	}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	validatingWebhook := admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.NamespacedScope, v.Operations())
+	return []admissionregistrationv1.ValidatingWebhook{*validatingWebhook}
+}
+
+// Admitters returns the admitter objects used to validate secrets.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+type admitter struct {
+	namespaceCache corev1controllers.NamespaceCache
+}
+
+// Admit handles the webhook admission request sent to this webhook.
+func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	listTrace := trace.New("project Admit", trace.Field{Key: "user", Value: request.UserInfo.Username})
+	defer listTrace.LogIfLong(admission.SlowTraceDuration)
+
+	oldProject, newProject, err := objectsv3.ProjectOldAndNewFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get old and new projects from request: %w", err)
+	}
+
+	if request.Operation == admissionv1.Delete {
+		return a.admitDelete(oldProject)
+	}
+	return a.admitCreateOrUpdate(oldProject, newProject)
+}
+
+func (a *admitter) admitDelete(project *v3.Project) (*admissionv1.AdmissionResponse, error) {
+	if project.Labels[systemProjectLabel] == "true" {
```

Comment:
Do we want to prevent users from removing this label in an update call as well? As it stands now, I think I could just remove the label and then my delete call would work (I don't think is is new or anything, more prudent).

Reply:
The norman store currently does not do this, so parity would mean that we should not add that here.

IMO this is just a guardrail, if someone really wants to remove the system project (maybe they accidentally created an extra somehow) then they can use this escape hatch to do it.

Reply:
Typically we don't add optional validations to the webhook - doing that pushes people toward doing things like overriding the webhook (which is a particularly dangerous operation). 

I think that this is justifiable since it's how norman behaves, but I think for a future enhancement we should either make this fully protected or remove it.